### PR TITLE
feat(stage) Add stage filmstrip (multiple participants on stage)

### PR DESCRIPTION
**Pull request by robertpin**
See multiple participants on stage
Pin and unpin to stage
Automatic selection of participants to be displayed on the stage filmstrip based on dominant speaker changes
Make Filmstrip a reusable component. Used by MainFilmstrip (old functionality) and the new StageFilmstrip
Rename DominantSpeakerName to StageParticipantNameLabel
Active border now showed only for the dominant speaker (no longer for the pinned participant)
Hide video from the vertical filmstrip for the participants on stage

### CODE REVIEW

**Comment by robertpin** [comment]
Jenkins test this please

File: react/features/filmstrip/actionTypes.js [hunk 1]
```diff
@@ -118,3 +118,44 @@ export const SET_USER_FILMSTRIP_WIDTH = 'SET_USER_FILMSTRIP_WIDTH';
...
+/**
+ * The type of Redux action which adds a participant to the active list
+ * (the participants displayed on the stage filmstrip).
+ * {
+ *     type: ADD_ACTIVE_PARTICIPANT,
+ *     participantId: string,
+ *     pinned: boolean
+ * }
+ */
+export const ADD_ACTIVE_PARTICIPANT = 'ADD_ACTIVE_PARTICIPANT';
```
Comment (quitrk) [inline react/features/filmstrip/actionTypes.js:141]: name is a bit misleading imo, maybe something like `ADD_STAGE_PARTICIPANT` along with the name changes for derived functionality..wdyt?

File: react/features/filmstrip/functions.web.js [hunk 2]
```diff
@@ -495,7 +498,7 @@ export function computeDisplayModeFromInput(input: Object) {
     } = input;
     const adjustedIsVideoPlayable = input.isVideoPlayable && (!isRemoteParticipant || canPlayEventReceived);
 
-    if (!tileViewActive && isScreenSharing && isRemoteParticipant) {
+    if ((!tileViewActive && isScreenSharing && isRemoteParticipant) || (isActiveParticipant && !tileViewActive)) {
```
Comment (quitrk) [inline react/features/filmstrip/functions.web.js:501]: `if (!tileViewActive && ((isScreenSharing && isRemoteParticipant) || isActiveParticipant))`?

File: react/features/filmstrip/components/web/ThumbnailWrapper.js [hunk 3]
```diff
@@ -117,25 +131,38 @@ class ThumbnailWrapper extends Component<Props> {
...
+        } else if (_verticalViewGrid) {
+            gridDimensions = gridView.gridDimensions;
+            thumbnailSize = gridView.thumbnailSize;
+        }
         const { columns, rows } = gridDimensions;
         const index = (rowIndex * columns) + columnIndex;
         let horizontalOffset;
         const { iAmRecorder } = state['features/base/config'];
-        const participantsLenght = remoteParticipantsLength + (iAmRecorder ? 0 : 1) - (disableSelfView ? 1 : 0);
+        const participantsLenght = stageFilmstrip ? remoteParticipantsLength
```
Comment (quitrk) [inline react/features/filmstrip/components/web/ThumbnailWrapper.js:164]: typo at `length`

**Review by quitrk — CHANGES_REQUESTED** [review]

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

File: react/features/filmstrip/functions.web.js [hunk 4]
```diff
@@ -642,3 +647,31 @@ export function isFilmstripScollVisible(state) {
...
+ * @param {Object} state - Redux state.
+ * @returns {boolean}
+ */
+export function shouldDisplayStageFilmstrip(state) {
+    const { activeParticipants } = state['features/filmstrip'];
+    const { remoteScreenShares } = state['features/video-layout'];
+    const { filmstrip } = state['features/base/config'];
+    const currentLayout = getCurrentLayout(state);
+
+    return !filmstrip?.disableStageFilmstrip && remoteScreenShares.length === 0
```
[the remark this answers is not in the document]
Comment (robertpin) [reply react/features/filmstrip/functions.web.js:675]: I don't think we want it to work with the horizontal filmstrip, but I'll check

File: react/features/conference/components/web/Conference.js [hunk 5]
```diff
@@ -55,7 +56,7 @@ const FULL_SCREEN_EVENTS = [
  * @private
  * @type {Object}
  */
-const LAYOUT_CLASSNAMES = {
+export const LAYOUT_CLASSNAMES = {
```
[the remark this answers is not in the document]
Comment (robertpin) [reply react/features/conference/components/web/Conference.js:59]: Yes, but implementing it like that is not trivial, so I'd rather do it as a refactor after this is merged

File: react/features/base/participants/functions.js [hunk 6]
```diff
@@ -290,8 +290,16 @@ export function getRemoteParticipantsSorted(stateful: Object | Function) {
...
  */
 export function getPinnedParticipant(stateful: Object | Function) {
-    const state = toState(stateful)['features/base/participants'];
-    const { pinnedParticipant } = state;
+    const state = toState(stateful);
+    const { pinnedParticipant } = state['features/base/participants'];
+    const { filmstrip } = state['features/base/config'];
+
+    if (!filmstrip?.disableStageFilmstrip) {
+        const { activeParticipants } = state['features/filmstrip'];
```
[the remark this answers is not in the document]
Comment (robertpin) [reply react/features/base/participants/functions.js:298]: This function is only called when there's just one (it's called in order to determine the participant on Large Video, which is displayed only when there's just one active participant)

File: config.js [hunk 7]
```diff
@@ -1303,6 +1303,10 @@ var config = {
     //     // Disables user resizable filmstrip. Also, allows configuration of the filmstrip
     //     // (width, tiles aspect ratios) through the interfaceConfig options.
     //     disableResizable: false,
+
+    //     // Disables the stage filmstrip
+    //     // (displaying multiple participants on stage besides the vertical filmstrip)
+    //     disableStageFilmstrip: false
```
[the remark this answers is not in the document]
Comment (robertpin) [reply config.js:1309]: I'll ask

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

File: react/features/filmstrip/functions.web.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (robertpin) [reply react/features/filmstrip/functions.web.js:675]: The answer is no. I'll change the code to make sure the feature is disabled for horizontal filmstrip

File: config.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (robertpin) [reply config.js:1310]: Yes we do

**Review by quitrk — CHANGES_REQUESTED** [review]

File: react/features/filmstrip/middleware.web.js [hunk 8]
```diff
@@ -82,12 +109,114 @@ MiddlewareRegistry.register(store => next => action => {
...
+        if (stageFilmstrip && currentLayout === LAYOUTS.VERTICAL_FILMSTRIP_VIEW) {
+            store.dispatch(addStageParticipant(id));
+        }
+        break;
+    }
+    case PARTICIPANT_LEFT: {
+        const { id } = action.participant;
+        const activeParticipantsIds = getActiveParticipantsIds(store.getState());
+
+        if (activeParticipantsIds.find(pId => pId === id)) {
```
[the remark this answers is not in the document]
Comment (robertpin) [reply react/features/filmstrip/middleware.web.js:214]: No, but we want the PARTICIPANT_LEFT to go through. The only thing we care about here is that if a stage participant leaves we clear the slot

[the remark this answers is not in the document]
Comment (quitrk) [reply react/features/filmstrip/middleware.web.js:214]: what I meant was, don't we just skip the check and always call remove?

File: react/features/filmstrip/components/web/StageFilmstrip.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    _videosClassName: string,
+
+    /**
+     * Whether or not the filmstrip videos should currently be displayed.
+     */
+    _visible: boolean
+};
+
+const StageFilmstrip = (props: Props) => props._currentLayout === LAYOUTS.VERTICAL_FILMSTRIP_VIEW && (
```
Comment (hristoterezov) [inline react/features/filmstrip/components/web/StageFilmstrip.js:95]: Shouldn't we check for stage filmstrip view here?

**Pull request merged**